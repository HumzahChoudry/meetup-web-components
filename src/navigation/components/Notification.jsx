import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import { FormattedRelative } from 'react-intl';

import Flex from '../../layout/Flex';
import FlexItem from '../../layout/FlexItem';
import Avatar from '../../media/Avatar';
import Icon from '../../media/Icon';



import { generateClassicUrl } from 'src/util/linksHelper';

const MEMBER_NOTIF_TYPES = ['mug_comm', 'convo', 'comment'];

/**
 * gets icon shape based on the "kind" of notification
 *
 * @param {String} notifKind - `kind` field of notification
 * @returns {String} - icon shape for notification
 */
export const getIconShape = notifKind => {
	let shape = 'meetup-m';

	const notifKindToShape = {
		convo: 'messages',
		mug_comm: 'messages',
		comment: 'messages',
		event: 'calendar',
		dues: 'creditcard',
		like: 'heart',
	};

	Object.keys(notifKindToShape).forEach(kind => {
		if (notifKind.startsWith(kind)) {
			shape = notifKindToShape[kind];
		}
	});

	return shape;
};

/**
 * @module Notification
 */
class Notification extends React.PureComponent {
	/**
	 * @constructor
	 * @param {Object} props passed in from parent
	 */
	constructor(props) {
		super(props);

		this.state = {
			isRead: props.isRead,
		};

		this.onClick = this.onClick.bind(this);
	}

	/**
	 * Marks notification read in internal state,
	 * calls fn to make POST to 'notifications/read' endpoint
	 *
	 * @return {undefined}
	 */
	onClick() {
		const { id, isRead, onMarkReadAction } = this.props;

		if (!isRead) {
			this.setState(() => ({ isRead: true }));
			onMarkReadAction(id);
		}
	}

	/**
	 * @return {React.element} rendered single notification
	 */
	render() {
		const {
			memberId,
			kind,
			updated,
			photoUrl,
			link,
			dangerouslySetInnerHTML,
			className,
			localeCode,
		} = this.props;

		const timeSince = new Date(updated);
		const isMemberRelated = MEMBER_NOTIF_TYPES.some(t => kind.startsWith(t));

		const classNames = cx(
			'list-item',
			'notification',
			{ 'notification--read': this.state.isRead },
			className
		);

		const avatarClassNames = cx('notification-image', {
			'avatar--person': isMemberRelated,
		});

		const notifUrlWithTracking = generateClassicUrl(
			localeCode,
			`/r/site/${kind}/${memberId}/notification/${link}`
		);

		return (
			<a href={notifUrlWithTracking} className={classNames} onClick={this.onClick}>
				<Flex>
					{photoUrl && (
						<FlexItem shrink className="valignChildren--center">
							<Avatar className={avatarClassNames} src={photoUrl} />
						</FlexItem>
					)}
					<FlexItem className="valignChildren--center">
						<h4
							className="text--small"
							dangerouslySetInnerHTML={dangerouslySetInnerHTML}
						/>
						<p className="text--hint text--small">
							<Icon shape={getIconShape(kind)} size="xs" />
							<time className="notification-time text--small">
								<FormattedRelative value={timeSince} />
							</time>
						</p>
					</FlexItem>
				</Flex>
			</a>
		);
	}
}

Notification.propTypes = {
	id: PropTypes.string,
	memberId: PropTypes.string,
	kind: PropTypes.string,
	isRead: PropTypes.bool.isRequired,
	onMarkReadAction: PropTypes.func.isRequired,
	link: PropTypes.string.isRequired,
	dangerouslySetInnerHTML: PropTypes.shape({
		__html: PropTypes.string,
	}).isRequired,
	photoUrl: PropTypes.string,
	updated: PropTypes.number,
	localeCode: PropTypes.number,
};

export default Notification;
