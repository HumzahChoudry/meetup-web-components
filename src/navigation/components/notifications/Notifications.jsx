import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import Flex from '../../../layout/Flex';
import FlexItem from '../../../layout/FlexItem';
import Avatar from '../../../media/Avatar';
import Icon from '../../../media/Icon';

const MEMBER_NOTIF_TYPES = ['mug_comm', 'convo', 'comment'];

/**
 * gets icon shape based on the "kind" of notification
 *
 * @param {String} notifKind - `kind` field of notification
 * @returns {String} - icon shape for notification
 */
export const getIconShape = notifKind => {
	let shape = 'meetup-m';

	const notifKindToShape = {
		convo: 'messages',
		mug_comm: 'messages',
		comment: 'messages',
		event: 'calendar',
		dues: 'creditcard',
		like: 'heart',
	};

	Object.keys(notifKindToShape).forEach(kind => {
		if (notifKind.startsWith(kind)) {
			shape = notifKindToShape[kind];
		}
	});

	return shape;
};

/**
 * @module Notification
 */
class Notification extends React.PureComponent {
	/**
	 * @constructor
	 * @param {Object} props passed in from parent
	 */
	constructor(props) {
		super(props);

		this.state = {
			isRead: props.isRead,
		};

		this.onClick = this.onClick.bind(this);
	}

	/**
	 * Marks notification read in internal state,
	 * calls fn to make POST to 'notifications/read' endpoint
	 *
	 * @return {undefined}
	 */
	onClick() {
		const { id, isRead, onMarkReadAction } = this.props;

		if (!isRead) {
			this.setState(() => ({ isRead: true }));
			onMarkReadAction(id);
		}
	}

	/**
	 * @return {React.element} rendered single notification
	 */
	render() {
		const {
			memberId,
			kind,
			formattedTimeSince,
			photoUrl,
			link,
			dangerouslySetInnerHTML,
			className,
			localeCode,
			generateClassicUrl,
		} = this.props;

		const isMemberRelated = MEMBER_NOTIF_TYPES.some(t => kind.startsWith(t));

		const classNames = cx(
			'list-item',
			'notification',
			{ 'notification--read': this.state.isRead },
			className
		);

		const avatarClassNames = cx('notification-image', {
			'avatar--person': isMemberRelated,
		});

		const notifUrlWithTracking = generateClassicUrl(
			localeCode,
			`/r/site/${kind}/${memberId}/notification/${link}`
		);

		return (
			<a href={notifUrlWithTracking} className={classNames} onClick={this.onClick}>
				<Flex>
					{photoUrl && (
						<FlexItem shrink className="valignChildren--center">
							<Avatar className={avatarClassNames} src={photoUrl} />
						</FlexItem>
					)}
					<FlexItem className="valignChildren--center">
						<h4
							className="text--small"
							dangerouslySetInnerHTML={dangerouslySetInnerHTML}
						/>
						<p className="text--hint text--small">
							<Icon shape={getIconShape(kind)} size="xs" />
							<time className="notification-time text--small">
								{formattedTimeSince}
							</time>
						</p>
					</FlexItem>
				</Flex>
			</a>
		);
	}
}

Notification.propTypes = {
	id: PropTypes.string,
	memberId: PropTypes.string,
	kind: PropTypes.string,
	isRead: PropTypes.bool,
	onMarkReadAction: PropTypes.func.isRequired,
	link: PropTypes.string,
	dangerouslySetInnerHTML: PropTypes.shape({
		__html: PropTypes.string,
	}),
	photoUrl: PropTypes.string,
	formattedTimeSince: PropTypes.string,
	localeCode: PropTypes.string.isRequired,
	className: PropTypes.string,
	notifUrlWithTracking: PropTypes.string,
	generateClassicUrl: PropTypes.func.isRequired,
};

export default Notification;
