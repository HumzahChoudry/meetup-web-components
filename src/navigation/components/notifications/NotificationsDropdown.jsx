// @flow
import React from 'react';

import Section from '../../../layout/Section';
import Chunk from '../../../layout/Chunk';

import Notification from './Notifications';

export const NotificationsDropdownComponent = ({
	self,
	notifications,
	onMarkReadAction,
	localeCode,
	emptyContentLabel,
	notificationTitleLabel,
	generateClassicUrl,
}) => {
	const isEmptyState = notifications.length < 1;

	const NotificationsList = () => (
		<div className="notifications--populated">
			<ul className="list">
				{notifications.map((notif, key) => (
					// TODO: refactor
					// eslint-disable-next-line react/no-array-index-key
					<li key={key}>
						<Notification
							className={key === 0 ? 'border--none' : ''}
							id={notif.id}
							memberId={self.id.toString()}
							kind={notif.kind}
							photoUrl={notif.photo ? notif.photo.photo_link : ''}
							isRead={notif.read}
							link={notif.link}
							onMarkReadAction={onMarkReadAction}
							dangerouslySetInnerHTML={{ __html: notif.text }}
							updated={notif.updated}
							localeCode={localeCode}
							generateClassicUrl={generateClassicUrl}
						/>
					</li>
				))}
			</ul>
		</div>
	);

	const NotificationsEmpty = () => (
		<Section className="notifications--empty align--center valignChildren--center">
			<Chunk>
				<p>{emptyContentLabel}</p>
			</Chunk>
		</Section>
	);

	return (
		<div className="align--left">
			<h2 className="padding--all border--bottom text--labelSecondary">
				{notificationTitleLabel}
			</h2>
			{isEmptyState ? <NotificationsEmpty /> : <NotificationsList />}
		</div>
	);
};

export default NotificationsDropdownComponent;
