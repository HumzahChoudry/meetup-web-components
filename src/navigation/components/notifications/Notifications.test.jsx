import React from 'react';
import { shallow } from 'enzyme';

import { MOCK_NOTIFICATION_EVENT } from 'meetup-web-mocks/lib/notifications/api';

import Notification, { getIconShape } from './Notifications';

describe('Notification Icons', () => {
	it('returns messages icon for message-related notifs', () => {
		expect(getIconShape('convo')).toBe('messages');
		expect(getIconShape('mug_comm')).toBe('messages');
		expect(getIconShape('comment')).toBe('messages');
	});
	it('returns event icon for event-related notifs', () => {
		expect(getIconShape('event')).toBe('calendar');
	});
	it('returns default icon for unspecified notif kinds', () => {
		expect(getIconShape('someOtherNotif')).toBe('meetup-m');
	});
});

describe('Notification component', () => {
	const MOCK_HANDLERS = {
		markRead: jest.fn(id => id),
	};

	const component = shallow(
		<Notification
			id={MOCK_NOTIFICATION_EVENT.id}
			memberId={MOCK_NOTIFICATION_EVENT.id}
			kind={MOCK_NOTIFICATION_EVENT.kind}
			isRead={false}
			onMarkReadAction={MOCK_HANDLERS.markRead}
			link={MOCK_NOTIFICATION_EVENT.link}
			dangerouslySetInnerHTML={{ __html: MOCK_NOTIFICATION_EVENT.text }}
			photoUrl={MOCK_NOTIFICATION_EVENT.photo.photo_link}
			updated={MOCK_NOTIFICATION_EVENT.updated}
			generateClassicUrl={jest.fn()}
		/>
	);

	it('should match snapshot', () => {
		expect(component).toMatchSnapshot();
	});
});
