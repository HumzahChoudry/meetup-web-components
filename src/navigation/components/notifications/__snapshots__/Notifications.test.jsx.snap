// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component should match snapshot 1`] = `
<a
  className="list-item notification"
  onClick={[Function]}
>
  <Flex
    direction="row"
  >
    <FlexItem
      className="valignChildren--center"
      shrink={true}
    >
      <Avatar
        className="notification-image"
        src="http://photos4.meetupstatic.com/photos/event/5/d/4/a/600_434003882.jpeg"
      />
    </FlexItem>
    <FlexItem
      className="valignChildren--center"
    >
      <h4
        className="text--small"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<b>Meetup Engineering</b> just announced a new Meetup for Apr 6. Only 13 spots.",
          }
        }
      />
      <p
        className="text--hint text--small"
      >
        <Icon
          shape="calendar"
          size="xs"
        />
        <time
          className="notification-time text--small"
        >
          <FormattedRelative
            updateInterval={10000}
            value={2017-03-30T15:46:22.000Z}
          />
        </time>
      </p>
    </FlexItem>
  </Flex>
</a>
`;
